core/anvil/src/region/mod.rs

```diff
@@ -31,62 +31,232 @@ const DATA_VERSION: i32 = 1631;
 /// Length, in bytes, of a sector.
 const SECTOR_BYTES: usize = 4096;
 
-/// The offset for each heightmap value
-const HEIGHTMAP_OFFSET: i64 = 9;
-
 /// Represents the data for a chunk after the "Chunk [x, y]" tag.
 #[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "PascalCase")]
 pub struct ChunkRoot {
-    #[serde(rename = "Level")]
     level: ChunkLevel,
-    #[serde(rename = "DataVersion")]
     data_version: i32,
 }
 
 /// Represents the level data for a chunk.
 #[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "PascalCase")]
 pub struct ChunkLevel {
     // TODO heightmaps, etc.
     #[serde(rename = "xPos")]
     x_pos: i32,
     #[serde(rename = "zPos")]
     z_pos: i32,
-    #[serde(rename = "Sections")]
     sections: Vec<LevelSection>,
-    #[serde(rename = "Biomes")]
     biomes: Vec<i32>,
-    #[serde(rename = "Entities")]
     entities: Vec<EntityData>,
     #[serde(rename = "TileEntities")]
     block_entities: Vec<BlockEntityData>,
-    #[serde(rename = "Heightmaps")]
-    heightmaps: Vec<i64>,
+    heightmaps: Heightmaps,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub struct Heightmaps {
+    // sometimes a few of those are missing, but we regenerate them anyways
+    #[serde(with = "packed_u9", default)]
+    light_blocking: Vec<u16>,
+    #[serde(with = "packed_u9", default)]
+    motion_blocking: Vec<u16>,
+    #[serde(with = "packed_u9", default)]
+    motion_blocking_no_leaves: Vec<u16>,
+    #[serde(with = "packed_u9", default)]
+    ocean_floor: Vec<u16>,
+    #[serde(with = "packed_u9", default)]
+    world_surface: Vec<u16>,
+}
+
+impl Heightmaps {
+    // TODO remove this and use hematite-nbt serde for serialization
+    pub(crate) fn pack_u9(u9_array: Vec<u16>) -> Vec<i64> {
+        let mut packed = Vec::with_capacity(36);
+        let mut iter = u9_array.iter();
+
+        let mut container = 0u64;
+        let mut shift = 0;
+        for _elem in 0..256 {
+            // For every element (u9)
+            let element = *iter.next().unwrap() as u64;
```

Comment:
We should return a `Result::Err` here rather than panicking. Right now, if we load in invalid heightmap data and then that chunk is saved, the server will panic. I'd prefer the chunk save process to fail with a log message.

Reply:
Yeah, I didn't put much effort into this method as I thought it to be temporary but I'll give it a once-over. Actually the server panics when it encounters a serialization error on saving regardless right now, because the Result is just being `expect`ed in the chunk save function

Reply:
Honestly, I'd rather just leave the panic in for now and remove this whole `Blob` business instead in the next PR. Switching everything to use the serde derived `Serialize` instead, now that array serialization is supported. Because right now everything in the chunk saving pipeline just panics when it encounters some error and I don't see the need to implement proper error propagation and handling just to scrap all of it in the next PR.